pxtlib/color.ts

```diff
@@ -0,0 +1,115 @@
+namespace pxt {
+    export function getWhiteContrastingBackground(color: string) {
+        if (contrastRatio("#ffffff", color) >= 4.5) return color;
+
+        const hslColor = hsl(color);
+
+        // There is probably a "smart" way to calculate this, but a cursory search
+        // didn't turn up anything so we're just going to decrease the luminosity
+        // until we get a contrasting color
+        const luminosityStep = 0.05;
+
+        let l = hslColor.l - luminosityStep;
+        while (l > 0) {
+            const newColor = hslToHex({ h: hslColor.h, s: hslColor.s, l });
+            if (contrastRatio("#ffffff", newColor) >= 4.5) return newColor;
+            l -= luminosityStep;
+        }
+
+        // We couldn't find one, so just return the original
+        console.warn(`Couldn't find a contrasting background for color ${color}`);
+        return color;
+    }
+
+    function hsl(color: string): { h: number, s: number, l: number } {
+        const rgb = pxt.toolbox.convertColor(color);
+
+        const r = parseInt(rgb.slice(1, 3), 16) / 255;
+        const g = parseInt(rgb.slice(3, 5), 16) / 255;
+        const b = parseInt(rgb.slice(5, 7), 16) / 255;
+
+        const max = Math.max(Math.max(r, g), b);
+        const min = Math.min(Math.min(r, g), b);
+
+        const diff = max - min;
+        let h;
+        if (diff === 0)
+            h = 0;
+        else if (max === r)
+            h = ((((g - b) / diff) % 6) + 6) % 6;
+        else if (max === g)
+            h = (b - r) / diff + 2;
+        else if (max === b)
+            h = (r - g) / diff + 4;
+
+        let l = (min + max) / 2;
+        let s = diff === 0
+            ? 0
+            : diff / (1 - Math.abs(2 * l - 1));
+
+        return {
+            h: h * 60,
+            s,
+            l
+        }
+    }
+
+
+    function relativeLuminance(color: string) {
+        const rgb = pxt.toolbox.convertColor(color);
+
+        const r = parseInt(rgb.slice(1, 3), 16) / 255;
+        const g = parseInt(rgb.slice(3, 5), 16) / 255;
+        const b = parseInt(rgb.slice(5, 7), 16) / 255;
+
+        const r2 = (r <= 0.03928) ? r / 12.92 : Math.pow((r + 0.055) / 1.055, 2.4);
+        const g2 = (g <= 0.03928) ? g / 12.92 : Math.pow((g + 0.055) / 1.055, 2.4);
+        const b2 = (b <= 0.03928) ? b / 12.92 : Math.pow((b + 0.055) / 1.055, 2.4);
+
+        return 0.2126 * r2 + 0.7152 * g2 + 0.0722 * b2;
+    }
+
+    function contrastRatio(fg: string, bg: string) {
+        return (relativeLuminance(fg) + 0.05) / (relativeLuminance(bg) + 0.05)
```

Comment:
Why do you need to add 0.05 to each of these values? Doesn't adding it to both the top and bottom cancel the effects?

Reply:
this is just the formula (I didn't come up with it), https://www.w3.org/TR/WCAG20/#contrast-ratiodef

adding to bottom and top doesn't cancel out, though! that's multiplication

Reply:
Ah, right. Thanks!